src/ucs/async/thread.c

```diff
@@ -171,42 +175,35 @@ static ucs_status_t ucs_async_thread_start(ucs_async_thread_t **thread_p)
         goto err_timerq_cleanup;
     }
 
-    /* Create epoll set the thread will wait on */
-    thread->epfd = epoll_create(1);
-    if (thread->epfd < 0) {
-        ucs_error("epoll_create() failed: %m");
-        status = UCS_ERR_IO_ERROR;
-        goto err_close_pipe;
+    status = ucs_event_set_create(&thread->event_set);
+    if (status != UCS_OK) {
+        goto err_free_event_set;
     }
 
-    /* Add wakeup pipe to epoll set */
+    /* Store file descriptor into void * storage without memory allocation. */
     wakeup_rfd    = ucs_async_pipe_rfd(&thread->wakeup);
-    memset(&event, 0, sizeof(event));
-    event.events  = EPOLLIN;
-    event.data.fd = wakeup_rfd;
-    ret = epoll_ctl(thread->epfd, EPOLL_CTL_ADD, wakeup_rfd, &event);
-    if (ret < 0) {
-        ucs_error("epoll_ctl(epfd=%d, ADD, fd=%d) failed: %m",thread->epfd,
-                  wakeup_rfd);
+    status = ucs_event_set_add(thread->event_set, wakeup_rfd,
+                               UCS_EVENT_SET_EVREAD,
+                               (void *)(uintptr_t)wakeup_rfd);
+    if (status != UCS_OK) {
+        ucs_error("ucs_event_set_add(fd=%d) failed: %m", wakeup_rfd);
         status = UCS_ERR_IO_ERROR;
-        goto err_close_epfd;
+        goto err_free_event_set;
     }
 
     ret = pthread_create(&thread->thread_id, NULL, ucs_async_thread_func, thread);
     if (ret != 0) {
         ucs_error("pthread_create() returned %d: %m", ret);
         status = UCS_ERR_IO_ERROR;
-        goto err_close_epfd;
+        goto err_free_event_set;
     }
 
     ucs_async_thread_global_context.thread = thread;
     status = UCS_OK;
     goto out_unlock;
 
-err_close_epfd:
-    close(thread->epfd);
-err_close_pipe:
```

Comment:
why it was removed?

Reply:
Sorry, It was wrong error handling fixed.

```diff
     status = ucs_event_set_create(&thread->event_set);
     if (status != UCS_OK) {
-        goto err_free_event_set;
+        goto err_close_pipe;
     }

/* snip */

 err_free_event_set:
     ucs_event_set_cleanup(thread->event_set);
+err_close_pipe:
+    ucs_async_pipe_destroy(&thread->wakeup);
 err_timerq_cleanup:
     ucs_timerq_cleanup(&thread->timerq);
```